File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = INT_MIN;
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return ans;
    }

    int dfs(TreeNode *t){
        if(t==NULL) return 0;
        int x = dfs(t->left);
        int y = dfs(t->right);
        ans = max({ans,x + t->val,y+t->val,x+y+t->val,t->val});
        return max(max(x,y) + t->val , t->val);
     }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        res = [root.val]
        def dfs(t):
            if not t:
                return 0
            
            l = max(dfs(t.left),0)
            r = max(dfs(t.right),0)
            res[0] = max(res[0],t.val + l + r)
            return t.val + max(l,r)
        dfs(root)
        return res[0]